core/assignment.go

```diff
@@ -149,38 +149,89 @@ func (c *StdAssignmentCoordinator) GetOperatorAssignment(state *OperatorState, q
 	return assignment, info, nil
 }
 
-func (c *StdAssignmentCoordinator) GetMinimumChunkLength(numOperators, blobLength, quantizationFactor uint, quorumThreshold, adversaryThreshold uint8) (uint, error) {
+func (c *StdAssignmentCoordinator) ValidateChunkLength(state *OperatorState, blobLength uint, info *BlobQuorumInfo) (bool, error) {
+
+	// Check that the chunk length meets the minimum requirement
+	if info.ChunkLength < MinChunkLength {
+		return false, ErrChunkLengthTooSmall
+	}
+
+	// Get minimum stake amont
+	minStake := state.Totals[info.QuorumID].Stake
+	for _, r := range state.Operators[info.QuorumID] {
+		if r.Stake.Cmp(minStake) < 0 {
+			minStake = r.Stake
+		}
+	}
+
+	totalStake := state.Totals[info.QuorumID].Stake
+	if info.ChunkLength != MinChunkLength {
+
+		num := new(big.Int).Mul(big.NewInt(2*int64(blobLength*percentMultiplier)), minStake)
+		denom := new(big.Int).Mul(big.NewInt(int64(info.QuorumThreshold-info.AdversaryThreshold)), totalStake)
+		maxChunkLength := uint(roundUpDivideBig(num, denom).Uint64())
+
+		maxChunkLength2 := roundUpDivide(2*blobLength*percentMultiplier, MaxRequiredNumChunks*uint(info.QuorumThreshold-info.AdversaryThreshold))
+
+		if maxChunkLength < maxChunkLength2 {
+			maxChunkLength = maxChunkLength2
+		}
+
+		maxChunkLength = uint(nextPowerOf2(uint64(maxChunkLength)))
+
+		if info.ChunkLength > maxChunkLength {
+			return false, ErrChunkLengthTooLarge
+		}
 
-	if adversaryThreshold >= quorumThreshold {
-		return 0, errors.New("invalid header: quorum threshold does not exceed adversary threshold")
 	}
 
-	numSys := roundUpDivide(uint(quorumThreshold-adversaryThreshold)*numOperators*quantizationFactor, PercentMultiplier)
-	chunkLength := roundUpDivide(blobLength, numSys)
-	return chunkLength, nil
+	return true, nil
 
 }
 
-func (c *StdAssignmentCoordinator) GetChunkLengthFromHeader(state *OperatorState, header *BlobQuorumInfo) (uint, error) {
+func (c *StdAssignmentCoordinator) CalculateChunkLength(state *OperatorState, blobLength, targetNumChunks uint, param *SecurityParam) (uint, error) {
+
+	chunkLength := uint(MinChunkLength) * 2
+
+	for {
```

Comment:
Oh, this is running a binary search to find the max C (as a power of 2), such that ValidateChunkLength(C) is true. The latent assumption here is that function ValidateChunkLength() is monotonic, which seems to be true.

Also I think the invariant here is that: when entering the for loop, `chunkLength/2` is a valid length.

Can we comment these out?

Related question: the reason we want to find the max C, is to minimize the num chunks (while satisfying the constraints)?

Reply:
> Oh, this is running a binary search
Not quite a binary search... its just going through the chunk lengths in order. But since only powers of 2 are accepted, the number of iterations will be logarithmic in max chunk length. 

> when entering the for loop, chunkLength/2 is a valid length.
Yes, because the initial chunk length is MinChunkLength*2, and MinChunkLength is always valid. 

> Can we comment these out?
Comment what out? 

> Related question: the reason we want to find the max C, is to minimize the num chunks (while satisfying the constraints)?
Yeah, that's the implicit reasoning here. The `targetNumChunks` provides an option to increase the num chunks if we want that for some reason.

Reply:
ok, good to know. I think it's useful to comment:
1) at the interface, comment this function will find the max chunk length (in power of 2) which satisfies the protocol constraints
2) at the implementation level here, comment that it's running binary search (btw it's actually binary search and it works only if Validate() function is monotoic. I think it's a variant of binary search called multiplicative binary search https://en.wikipedia.org/wiki/Multiplicative_binary_search)